nsx/resource_nsx_edge_dlr.go

```diff
@@ -0,0 +1,232 @@
+							Required:     true,
+							ForceNew:     false,
+							ValidateFunc: validateIP,
+						},
+						"mask": &schema.Schema{
+							Type:         schema.TypeString,
+							Required:     true,
+							ForceNew:     false,
+							ValidateFunc: validateIP,
+						},
+						"logical_switch_id": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: false,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceNsxEdgeDLRInterfaceCreate(d *schema.ResourceData, meta interface{}) error {
+	dlr, err := parseAndValidateDLRResourceData(d, meta)
+	if err != nil {
+		log.Printf("[ERROR] Configuration validation failed.")
+		return err
+	}
+
+	log.Printf("[INFO] Adding DLR Interface '%#v' to Edge '%s'", dlr, dlr.edgeId)
+
+	client := meta.(*govnsx.Client)
+	dlrInterfaces := nsxresource.NewEdgeDLRInterfaces(client)
+
+	edgeId := dlr.edgeId
+	ifaces := []nsxtypes.EdgeDLRInterface{}
+
+	for _, ifcfg := range dlr.ifCfgList {
+
+		addrGroups := []nsxtypes.AddressGroup{nsxtypes.AddressGroup{
+			PrimaryAddress: ifcfg.ip,
+			SubnetMask:     ifcfg.mask}}
+
+		iface := nsxtypes.EdgeDLRInterface{
+			AddressGroups: addrGroups,
+			Name:          ifcfg.name,
+			ConnectedToId: ifcfg.logical_switch_id,
+			Type:          "internal",
+			IsConnected:   true}
+
+		ifaces = append(ifaces, iface)
+	}
+
+	addInterfacesSpec := &nsxtypes.EdgeDLRAddInterfacesSpec{
+		EdgeDLRInterfaceList: ifaces,
+	}
+
+	_, err = dlrInterfaces.Post(addInterfacesSpec, edgeId)
+	if err != nil {
+		log.Printf("[ERROR] dlrInterfaces.Post () returned error : %v", err)
+		return err
+	}
+
+	d.SetId(fmt.Sprintf(DLRResourceIdPrefix + dlr.edgeId))
+
+	return resourceNsxEdgeDLRInterfaceRead(d, meta)
+}
+
+func resourceNsxEdgeDLRInterfaceRead(d *schema.ResourceData, meta interface{}) error {
+
+	client := meta.(*govnsx.Client)
+	dlrInterfaces := nsxresource.NewEdgeDLRInterfaces(client)
+
+	edgeId := d.Get("edge_id").(string)
+
+	log.Printf("[INFO] Read NSX Edge Router Interface: ", edgeId)
+	_, err := dlrInterfaces.Get(edgeId)
+
+	if err != nil {
+		log.Printf("[ERROR] Retriving Edge '%s' failed with error : '%v'", edgeId, err)
+		d.SetId("")
+		return err
+	}
+	return nil
+}
+
+func resourceNsxEdgeDLRInterfaceUpdate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[INFO] Update NSX Edge Router Interface TBD")
+	return nil
+}
+
+func resourceNsxEdgeDLRInterfaceDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*govnsx.Client)
+	iface := nsxresource.NewEdgeDLRInterfaces(client)
+
+	edgeId := d.Get("edge_id").(string)
+	log.Printf("[INFO] Deleting NSX EdgeInterface:%s %s\n", edgeId)
+
+	err := iface.Delete(edgeId)
+	if err != nil {
+		log.Printf("[Error] NSX Edge Interface Delete returned error : %v", err)
+		return err
+	}
+	return nil
+}
+
+func parseAndValidateDLRResourceData(d *schema.ResourceData, meta interface{}) (*dlrCfg, error) {
+
+	edgeType, err := getEdgeType(d.Get("edge_id").(string), meta)
```

Comment:
Can we refactor the edge type check in create and back populating the state in Read